Orange/widgets/visualize/owscatterplotgraph.py

```diff
@@ -598,480 +491,308 @@ def update_tooltip(self, modifiers):
+        self.view_box.init_history()
+        self.view_box.tag_history()
+
+    def get_coordinates(self):
+        x, y = self.master.get_coordinates_data()
+        if x is None:
             self.n_points = 0
-            self.master.Warning.missing_coords(
-                self.shown_x.name, self.shown_y.name)
+            return None, None
+        x, y = self.jitter_coordinates(x, y)
+        self.n_points = len(x) if x is not None else 0
+        return x, y
+
+    def jitter_coordinates(self, x, y):
+        if self.jitter_size == 0:
+            return x, y
+        return self._jitter_data(x, 0), self._jitter_data(y, 1)
+
+    @classmethod
+    def _update_plot_coordinates(cls, plot, x, y):
+        data = dict(x=x, y=y)
+        for prop in ('pen', 'brush', 'size', 'symbol', 'data'):
+            data[prop] = plot.data[prop]
+        plot.setData(**data)
+
+    def update_coordinates(self):
+        x, y = self.get_coordinates()
+        if x is None or not len(x):
             return
+        if self.scatterplot_item is None:
+            kwargs = {"x": x, "y": y, "data": np.arange(self.n_points)}
+            self.scatterplot_item = ScatterPlotItem(**kwargs)
+            self.scatterplot_item.sigClicked.connect(self.select_by_click)
+            self.scatterplot_item_sel = ScatterPlotItem(**kwargs)
+            self.plot_widget.addItem(self.scatterplot_item_sel)
+            self.plot_widget.addItem(self.scatterplot_item)
+            self.update_point_props()
+        else:
+            self._update_plot_coordinates(self.scatterplot_item, x, y)
+            self._update_plot_coordinates(self.scatterplot_item_sel, x, y)
 
-        self.x_data, self.y_data = self.get_xy_data_positions(
-            attr_x, attr_y, self.valid_data)
-        self.n_points = len(self.x_data)
-
-        if reset_view:
-            min_x, max_x = np.nanmin(self.x_data), np.nanmax(self.x_data)
-            min_y, max_y = np.nanmin(self.y_data), np.nanmax(self.y_data)
-            self.view_box.setRange(
-                QRectF(min_x, min_y, max_x - min_x, max_y - min_y),
-                padding=0.025)
-            self.view_box.init_history()
-            self.view_box.tag_history()
-        [min_x, max_x], [min_y, max_y] = self.view_box.viewRange()
-
-        for axis, var in (("bottom", attr_x), ("left", attr_y)):
-            self.set_axis_title(axis, var)
-            if var.is_discrete:
-                self.set_labels(axis, get_variable_values_sorted(var))
-            else:
-                self.set_labels(axis, None)
+        self.update_label_coords(x, y)
+        self.update_density()  # Todo: doesn't work: try MDS with density on
+        self._reset_view(x, y)
 
-        # compute overlaps of points for use in compute_colors and compute_sizes
-        self.overlaps = []
         self.coord_to_id = defaultdict(list)
-        for i, xy in enumerate(zip(self.x_data, self.y_data)):
+        for i, xy in enumerate(zip(x, y)):
```

Comment:
This code looks expensive, but it's needed only when sizing and coloring by overlaps.

Plus, I believe it belong to the widget, not the plot (see other comments).